gdal/port/cpl_vsil_s3.cpp

```diff
@@ -241,32 +343,211 @@ void VSICurlFilesystemHandler::AnalyseS3FileList(
+                {
+                    poS3FS->UpdateMapFromHandle(poS3HandleHelper);
                 }
+                CPLFree(sWriteFuncData.pBuffer);
+                CPLFree(sWriteFuncHeaderData.pBuffer);
+            }
+            else
+            {
+                CPLDebug(poS3FS->GetDebugKey(), "%s",
+                         sWriteFuncData.pBuffer
+                         ? sWriteFuncData.pBuffer : "(null)");
+                CPLFree(sWriteFuncData.pBuffer);
+                CPLFree(sWriteFuncHeaderData.pBuffer);
+                curl_easy_cleanup(hCurlHandle);
+                return false;
             }
         }
+        else
+        {
+            const bool bIgnoreGlacier = CPLTestBool(
+                CPLGetConfigOption("CPL_VSIL_CURL_IGNORE_GLACIER_STORAGE", "YES"));
+            bool bIsTruncated;
+            bool ret = AnalyseS3FileList( osBaseURL,
+                                          sWriteFuncData.pBuffer,
+                                          bIgnoreGlacier,
+                                          bIsTruncated );
 
-        if( osFileList.size() == 0 )
+            CPLFree(sWriteFuncData.pBuffer);
+            CPLFree(sWriteFuncHeaderData.pBuffer);
+
+            curl_easy_cleanup(hCurlHandle);
+            return ret;
+        }
+
+        curl_easy_cleanup(hCurlHandle);
+    }
+}
+
+/************************************************************************/
+/*                           NextDirEntry()                             */
+/************************************************************************/
+
+const VSIDIREntry* VSIDIRS3::NextDirEntry()
+{
+    while( true )
+    {
+        if( nPos < static_cast<int>(aoEntries.size()) )
+        {
+            auto& entry = aoEntries[nPos];
+            nPos ++;
+            return entry.get();
+        }
+        if( osNextMarker.empty() )
+        {
+            return nullptr;
+        }
+        if( !IssueListDir() )
         {
-            // To avoid an error to be reported
-            osFileList.AddString(".");
+            return nullptr;
         }
     }
+}
 
-    CPLDestroyXMLNode(psTree);
+/************************************************************************/
+/*                          AnalyseS3FileList()                         */
+/************************************************************************/
+
+void VSICurlFilesystemHandler::AnalyseS3FileList(
+    const CPLString& osBaseURL,
+    const char* pszXML,
+    CPLStringList& osFileList,
+    int nMaxFiles,
+    bool bIgnoreGlacierStorageClass,
+    bool& bIsTruncated )
+{
+    VSIDIRS3 oDir(this);
+    oDir.nMaxFiles = nMaxFiles;
+    bool ret =
+        oDir.AnalyseS3FileList(osBaseURL, pszXML,
+                               bIgnoreGlacierStorageClass, bIsTruncated);
+    for(const auto &entry: oDir.aoEntries )
+    {
+        osFileList.AddString(entry->pszName);
+    }
+    if( ret && osFileList.empty() )
+    {
+        osFileList.AddString(".");
```

Comment:
We seem to be stripping `.` & `..` above but we're adding it in here?

Reply:
That's for the purpose of the historical ReadDir() interface which returns a char**. If you don't have an entry, then NULL is returned and it is not clear if it means an error occured (the directory doesn't exist) or if it is empty

Reply:
So should `ReadDir()` be adding that in rather than the VSI drivers themselves?

Reply:
ReadDir() is a virtual method of the VSIFilesystemHandler class, so only the implementations know if the directory doesn't exist or is empty. That said, the 'cloud drivers' implement a GetFileList() method that is used by the generic VSICurlFilesystemHandler code and has a success flag, so I've changed the cloud drivers to not add themselves the '.' entry but let the generic code do it if needed

One potential solution to avoid adding the '.' would be consider char ** = NULL as en error, and char ** = array with a NULL terminated pointer as an empty directory. That could work in the C/C++ side (although CSLDuplicate used in a few places consider both the same), and I'm not sure how that would be translated in Python. We probably don't want to distinguish None from []

Reply:
Python's stdlib generally follows:
```
In [2]: os.listdir('/tmp/exists/')
Out[2]: []

In [3]: os.listdir('/tmp/does-not-exist/')
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
<ipython-input-3-7c6192b2cfc3> in <module>()
----> 1 os.listdir('/tmp/does-not-exist/')

FileNotFoundError: [Errno 2] No such file or directory: '/tmp/does-not-exist/'

In [5]: list(os.walk('/tmp/does-not-exist/'))
Out[5]: []

In [6]: list(os.walk('/tmp/exists/'))
Out[6]: [('/tmp/exists/', [], [])]

In [7]: list(scandir.scandir('/tmp/exists/'))
Out[7]: []

In [8]: list(scandir.scandir('/tmp/does-not-exist/'))
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
<ipython-input-8-cbedeeea6471> in <module>()
----> 1 list(scandir.scandir('/tmp/does-not-exist/'))

OSError: [Errno 2] No such file or directory: '/tmp/does-not-exist/'
```

Proving me wrong about consistency 😉